syntax = "proto3";


// ######################
// ###### REQUESTS ###### 
// ######################
message IdRequest {
    string user_id = 1;
}

message UsernameRequest{
    string user_name = 1;
}

message CreateUserRequest{
    string user_name = 1;
}

message GetUsersRequest{
    int32 max_result = 1;
}

// ######################
// ##### RESPONSES ###### 
// ######################
message UserData {
    string user_name = 1;
    int32 user_num_reviews = 2;
    int32 user_num_games_owned = 3;
    string user_id = 4;
}

message UsersDataList {
    repeated UserData users = 1;
}

message CreateUserResponse{
    string user_name = 1;
    string user_id = 2;
    int32 user_num_games_owned = 3;
    int32 user_num_reviews = 4;
    int32 user_playtime_forever = 5;
    int32 user_playtime_last_two_weeks = 6;
    int32 user_playtime_at_review = 7;
    int32 author_last_played = 8;
    string user_pwd = 9;
}

message DeletionResponse{
    bool deleted = 1;
}


// ######################
// ##### SERVICES #######
// ######################

service Users {
    rpc GetUserById (IdRequest) returns (UserData);
    rpc GetUserByUsername (UsernameRequest) returns (UserData);
    rpc CreateUser (CreateUserRequest) returns (CreateUserResponse);
    rpc GetUsers (GetUsersRequest) returns (UsersDataList);
    rpc DeleteUserByUsername (UsernameRequest) returns (DeletionResponse);
    rpc DeleteUserById (IdRequest) returns (DeletionResponse);
}